pkg/controller/keyvault/spc/reconcile.go

```diff
@@ -0,0 +1,265 @@
+	modifyOwner func(obj client.Object) error
+}
+
+type secretProviderClassReconciler[objectType client.Object] struct {
+	// config options
+	name controllername.ControllerNamer
+	// toSpcNamer is a function that returns an iterator for each SecretProviderClass that should be managed for the given object
+	toSpcOpts func(context.Context, client.Client, objectType) iter.Seq2[spcOpts, error]
+
+	// set during constructor
+	client client.Client
+	events record.EventRecorder
+	config *config.Config
+}
+
+func (s *secretProviderClassReconciler[objectType]) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, retErr error) {
+	defer func() {
+		metrics.HandleControllerReconcileMetrics(s.name, result, retErr)
+	}()
+
+	logger, err := logr.FromContext(ctx)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("getting logger from context: %w", err)
+	}
+	logger = s.name.AddToLogger(logger).WithValues("name", req.Name, "namespace", req.Namespace)
+
+	logger.Info("getting Object")
+	obj := util.NewObject[objectType]()
+	if err := s.client.Get(ctx, req.NamespacedName, obj); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+
+		return ctrl.Result{}, fmt.Errorf("getting object: %w", err)
+	}
+	logger = logger.WithValues("generation", obj.GetGeneration())
+
+	objUpdated := false
+	for spcOpts, err := range s.toSpcOpts(ctx, s.client, obj) {
+		if err != nil {
+			var userErr util.UserError
+			if errors.As(err, &userErr) {
+				logger.Info(fmt.Sprintf("failed to build secret provider class with user error: %s", userErr.Error()))
+				s.events.Eventf(obj, corev1.EventTypeWarning, "InvalidInput", "error while processing Keyvault reference: %s", userErr.UserError())
+				return ctrl.Result{}, nil
+			}
+
+			logger.Error(err, "failed to build secret provider class")
+			return ctrl.Result{}, fmt.Errorf("building secret provider class: %w", err)
+		}
+
+		if spcOpts.action == actionCleanup {
+			if err := s.cleanupSpc(ctx, logger, spcOpts); err != nil {
+				logger.Error(err, "failed to clean up SecretProviderClass")
+				return ctrl.Result{}, fmt.Errorf("cleaning up SecretProviderClass: %w", err)
+			}
+
+			continue
+		}
+
+		spc, err := s.buildSpc(obj, spcOpts)
+		if err != nil {
+			logger.Error(err, "failed to build SecretProviderClass spec")
+			return ctrl.Result{}, fmt.Errorf("building SecretProviderClass spec: %w", err)
+		}
+		logger = logger.WithValues("spc", spc.Name)
+
+		logger.Info("reconciling SecretProviderClass")
+		if err := util.Upsert(ctx, s.client, spc); err != nil {
+			err := fmt.Errorf("failed to reconcile SecretProviderClass %s: %w", spc.Name, err)
+			s.events.Eventf(obj, corev1.EventTypeWarning, "FailedUpdateOrCreateSPC", "error while creating or updating SecretProviderClass needed to pull Keyvault reference: %s", err.Error())
+			logger.Error(err, "failed to upsert SecretProviderClass")
+			return ctrl.Result{}, err
+		}
+
+		if spcOpts.modifyOwner != nil {
```

Comment:
why not include the objUpdated clause in here too?
```suggestion
		if spcOpts.modifyOwner != nil {
				if err := spcOpts.modifyOwner(obj); err != nil {
					logger.Error(err, "failed to modify owning object")
					return ctrl.Result{}, fmt.Errorf("modifying owning object: %w", err)
				}
						logger.Info("updating owning object ")
				if err := s.client.Update(ctx, obj); err != nil {
				...
		}
				
```

Reply:
because sometimes we have to update the owner multiple times and it doesn't make sense to call api server update multiple times